lib/improver/spotdata_new/apply_lapse_rate.py

```diff
@@ -0,0 +1,206 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# (C) British Crown Copyright 2017-2018 Met Office.
+# All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# * Redistributions of source code must retain the above copyright notice, this
+#   list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright notice,
+#   this list of conditions and the following disclaimer in the documentation
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Apply temperature lapse rate adjustments to a spot data cube."""
+
+import numpy as np
+
+import iris
+from improver.spotdata_new.spot_extraction import SpotExtraction
+
+
+class SpotLapseRateAdjust:
+    """
+    Adjusts spot data temperatures by a lapse rate to better represent the
+    conditions at their altitude that may not be captured by the model
+    orography.
+    """
+
+    def __init__(self, neighbour_selection_method='nearest',
+                 grid_metadata_identifier='mosg'):
```

Comment:
You might find this identifier causes problems as StaGE attributes are added / modified, eg does `mosg__forecast_run_duration` need to match?  Tempted to suggest requiring a list of grid-defining attribute names unless we're confident of this.  Definitely suggest setting this as required or "None" default if we don't want unexpected failures (the CLI default can of course override this).

Reply:
I'm wary of making it None, as that means it will perform no check at all, though it would remove the Met Office specific default here. We are working with indices, so the code will blithely return data which may be nowhere near the spot-data site unless we have performed some kind of check; the only sign of a problem would be if the neighbour indices went beyond the bounds of the diagnostic grid. This all suggests that it is a very reasonable idea to make it a mandatory argument, so I will do that.

We could make this into a list, but for our uses we will want probably to always check the ```mosg__grid``` attributes. This would confirm the domain (```mosg__grid_domain```), the grid type (```mosg__grid_type```), and the version (```mosg__grid_version```). This should capture anything relating to grid changes that we would need to be aware of (e.g. a version change to 2.0 that doubles the resolution say). This is also how the NeighbourSelection plugin was written. It copies attributes into the neighbour cube that match a ```grid_metadata_identifier```, currently set by default to ```mosg```, so this is the metadata available to compare in the neighbour cube.